packages/drivers/odsp-driver/src/fetchWithEpochValidation.ts

```diff
@@ -0,0 +1,125 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import assert from "assert";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { createOdspNetworkError, fluidEpochMismatchError, OdspErrorType } from "@fluidframework/odsp-doclib-utils";
+import { fetchAndParseAsJSONHelper, fetchHelper, IOdspResponse } from "./odspUtils";
+import { ICacheEntry, IPersistedCache, IPersistedCacheValue } from "./odspCache";
+
+/**
+ * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
+ * server can match it with its epoch value in order to match the version.
+ * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
+ * then it also clears all the cached entries for the given container.
+ */
+export class FetchWithEpochValidation {
+    private _fluidEpoch: string | undefined;
+    private _hashedDocumentId: string | undefined;
+    constructor(
+        private readonly persistedCache: IPersistedCache,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
+        this._hashedDocumentId = docId;
+    }
+
+    public get hashedDocumentId(): string | undefined {
+        return this._hashedDocumentId;
+    }
+
+    public get fluidEpoch() {
+        return this._fluidEpoch;
+    }
+
+    public async fetchFromCache<T>(entry: ICacheEntry, maxOpCount: number | undefined): Promise<T | undefined> {
+        const value: IPersistedCacheValue = await this.persistedCache.get(entry, maxOpCount);
+        if (value !== undefined) {
+            try {
+                this.validateEpochFromResponse(value.fluidEpoch);
+            } catch (error) {
+                this.checkForEpochError(createOdspNetworkError("Epoch Mismatch", fluidEpochMismatchError));
```

Comment:
I think a bit cleaner would be to have a function called cleanCacheEntries() and calling it here, instead of reusing server error code. That way fluidEpochMismatchError does not need to exist- appropriate error code can be directly references in switch statement, making it way more clear that this code is used by server (I initially thought this is client-only code)

Reply:
I have added proper description for this: fluidEpochMismatchError  which makes sure that this is thrown by server. If I replace "checkForEpochError" with "cleanCacheEntries" then I need to check whether the error is epoch mismatch error in all fetch apis instead at one place. So I find this way better. Plus it does not make much sense to keep both because then we also want to send logger event which is done in "checkForEpochError". So I think better is to add proper description.